lib/gcov/ukgcov.c

```diff
@@ -0,0 +1,203 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ * Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the MIT License (the "License", see COPYING.md).
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <string.h>
+#include <uk/essentials.h>
+#include <uk/gcov.h>
+#include <uk/config.h>
+
+#ifdef CONFIG_LIBGCOV_OUTPUT_BINARY_MEMORY
+static unsigned char __used *gcov_output_buffer;
+static size_t __used gcov_output_buffer_pos;
+static size_t gcov_output_buffer_size = INITIAL_BUFFER_SIZE;
+#endif /* CONFIG_LIBGCOV_OUTPUT_BINARY_MEMORY */
+
+/* gcov declarations taken from
+ * https://github.com/gcc-mirror/gcc/blob/master/libgcc/gcov.h
+ */
+struct gcov_info;
+typedef void (*__filename_fn)(const char *, void *);
+typedef void (*__dump_fn)(const void *, unsigned int, void *);
+typedef void *(*__allocate_fn)(unsigned int, void *);
+
+void __gcov_info_to_gcda(const struct gcov_info *__info,
+			 __filename_fn filename_fn, __dump_fn dump_fn,
+			 __allocate_fn allocate_fn, void *__arg);
+void __gcov_filename_to_gcfn(const char *__filename, __dump_fn dump_fn,
+			     void *__arg);
+
+#ifdef CONFIG_LIBGCOV_OUTPUT_BINARY_MEMORY
+static void dump_memory(const void *source, unsigned int n, void *arg __unused)
+{
+	const unsigned char *c = source;
+	unsigned char *buff;
```

Comment:
Cool kids use `buf` 😛 It's also consistent with the other functions in this file that use `buf` too.